### PR TITLE
Initial public draft for TotW 130

### PR DESCRIPTION
**Pull request by tituswinters**

### CODE REVIEW

File: _posts/2017-11-30-totw-130.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+sidenav: side-nav-tips.html
+published: false
+permalink: tips/130
+type: markdown
+order: "130"
+---
+
+Originally posted as totw/130 on 2017-02-17
+
+By Titus Winters (titus@google.com)
```
Comment (manshreck) [inline _posts/2017-11-30-totw-130.md]: Linkify email address.

File: _posts/2017-11-30-totw-130.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+*The precision of naming takes away from the uniqueness of seeing* — Pierre
+Bonnard
```
Comment (manshreck) [inline _posts/2017-11-30-totw-130.md]: Include author in italics quote

File: _posts/2017-11-30-totw-130.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+The earliest commit of the Google C++ Style Guide contains the guidance that
+many people are still using for namespace naming. Roughly, this can be
+summarized as "namespaces are derived from package paths," although there's an
+interesting less-followed option that an entire top-level package
+(search/...) could choose a single flat top-level namespace. Following on
```
Comment (manshreck) [inline _posts/2017-11-30-totw-130.md]: Omit search/ or genericize ?

In either case, enclose in \` \`

File: _posts/2017-11-30-totw-130.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+the heels of Java's package naming requirements, this makes a lot of sense: we
+want to be able to uniquely identify symbols in C++ and we want there to be
+uniqueness and consistency in namespace choice.
+
+Except in actuality, we don't. We just didn't realize for almost a decade.
+
+## Name lookup
```
Comment (manshreck) [inline _posts/2017-11-30-totw-130.md]: initial caps headings

File: _posts/2017-11-30-totw-130.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+Let's start with how name lookup works in C++ and how it's different from Java.
+
+``` c++
+namespace foo {
+namespace bar {
+void f() {
```
Comment (manshreck) [inline _posts/2017-11-30-totw-130.md]: Mildly prefer a blank line between namespace and f(). Don't know if this is a style violation :)

File: _posts/2017-11-30-totw-130.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+  Baz b;
+}
+}
+}
+```
+
+In C++, lookup on an unqualified name (`Baz`) will search expanding scopes for
```
Comment (manshreck) [inline _posts/2017-11-30-totw-130.md]: f(), bar(), foo(), etc. throughout

File: _posts/2017-11-30-totw-130.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+``` java
+import com.google.foo.bar.Baz;
+import com.google.foo.bar.*;
+```
+
+In no case is `Baz` looked for outside of the package that is explicitly
+provided: wildcards don't descend into child packages, nor is search extended
+into parent packages. As it turns out, the difference in how parent
+packages/namespaces is handled is fundamental to why structural namespace naming
+(making the namespace structure match the package hierarchy) is a mistake.
```
Comment (manshreck) [inline _posts/2017-11-30-totw-130.md]: As it turns out, **this** difference ...  is a mistake **(in C++)** (?)

File: _posts/2017-11-30-totw-130.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+}
+```
+
+In C++, lookup on an unqualified name (`Baz`) will search expanding scopes for
+a symbol of the same name: first in `f` (the function), then in `bar`, then in
+`foo`, then in the global namespace.
+
+
+In Java, there is no such thing as an unqualified symbol: either a symbol is
+a qualified name
```
Comment (manshreck) [inline _posts/2017-11-30-totw-130.md]: name -> name:

File: _posts/2017-11-30-totw-130.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+``` java
+public void f() {
+  com.google.foo.bar.Baz b = new com.google.foo.bar.Baz();
+}
+```
+
+or it is imported, either as a single package member or via wildcard:
```
Comment (manshreck) [inline _posts/2017-11-30-totw-130.md]: It's OK to capitalize Or in this case

File: _posts/2017-11-30-totw-130.md [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+And what's worse: unqualified search starts at the bottom of that list *and
+stops as soon as there is a namespace match*. This means that your build can be
+broken if any of your transitive includes add a previously unused namespace that
+matches the leading namespace of any symbol you use out of an unqualified
+namespace. Strictly speaking, this doesn't even have to be a build break: if
+someone adds something with a matching name and a syntactically-compatible API,
+the implementation of that API may be completely incompatible and cause
+widespread havoc at runtime. Obviously this isn't too bad with `std` - nobody
+should ever be adding a nested namespace `std`, but what about more common
```
Comment (manshreck) [inline _posts/2017-11-30-totw-130.md]: \`std\`, -> \`std\`  -

(You introduce the aside with a dash, so you need to close it with one)

File: _posts/2017-11-30-totw-130.md [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+sub-namespaces. This is a recipe for unnecessary and unintended collisions.
+
+For comparison, the problem in Java is far reduced: if you wildcard-import from
+two packages in Java and one adds a new symbol with the same name as the other
+package, your build can break. This is easily and completely solved by
+forbidding wildcard imports as is done in many Java styles.
+
+## Two Consistent Options, Three Approaches
+
+There are two features that prevent this build-break-at-a-distance.
```
Comment (manshreck) [inline _posts/2017-11-30-totw-130.md]: distance:

File: _posts/2017-11-30-totw-130.md [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+
+*   If no leaf namespace (`search::foo::bar`) matches any top-level namespace
+    (`::bar`) or a sub-namespace of any parent of that leaf (`search::bar`), there
```
Comment (manshreck) [inline _posts/2017-11-30-totw-130.md]: Nit: "there will be no problems" sounds a bit awkward, but i can live with it.

"no name collisions will occur" (?)

File: _posts/2017-11-30-totw-130.md [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+    will be no problems.
+*   If there are no unqualified lookups, there will be no problems.
+
+There are (at least) three ways to accomplish these:
```
Comment (manshreck) [inline _posts/2017-11-30-totw-130.md]: these: -> the behavior as noted above:

File: _posts/2017-11-30-totw-130.md [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+
+*   Always fully qualify everything outside of the current namespace. This is
+    very verbose and sort of weird: nothing in C++ (including the standard
+    library) is written with leading `::` on every symbol. 
+*   Build some tooling to identify introduction of new namespaces and ensure
+    that it doesn't overlap with any other namespace in the same hierarchy. That
+    is, no adding `search::bar` if there is a `::bar` or an `search::foo::bar`.
```
Comment (manshreck) [inline _posts/2017-11-30-totw-130.md]: no adding -> don't add

Comment (manshreck) [inline _posts/2017-11-30-totw-130.md]: an \`search::foo::bar\` -> a \`search:;foo::bar\`

File: _posts/2017-11-30-totw-130.md [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+*   Don't nest deeply: a single top-level namespace per project gets the same
+    result without long/complicated names, with less exposure to accidents,
+    without causing surprise for new engineers, and without the need to build any
+    tooling.
+
+The current style guide suggests the [last
+option](http://google.github.io/styleguide/cppguide.html#Namespace_Names), but
+allows for the old style (namespaces match package names) if necessary. This is
+largely because the style arbiters don't want to cause too much anxiety or
```
Comment (manshreck) [inline _posts/2017-11-30-totw-130.md]: the style arbiters don't -> Google didn't (?)

File: _posts/2017-11-30-totw-130.md (hunk 2 above)
Comment (tituswinters) [reply _posts/2017-11-30-totw-130.md]: I don't think we've been doing that. You sure?

File: _posts/2017-11-30-totw-130.md (hunk 3 above)
Comment (tituswinters) [reply _posts/2017-11-30-totw-130.md]: just dropping it, it's less relevant externally

File: _posts/2017-11-30-totw-130.md (hunk 5 above)
Comment (tituswinters) [reply _posts/2017-11-30-totw-130.md]: yeah, I think that'd be a violation

File: _posts/2017-11-30-totw-130.md (hunk 7 above)
Comment (tituswinters) [reply _posts/2017-11-30-totw-130.md]: Hmm. It's an intricate sentence.

as it turns out
(the difference in how parent packages/namespaces are handled) is fundamental to
(why structural namespace naming (...)) is a mistake.

Suggestions welcome.

**Comment by tituswinters** [comment]
Updated, PTAL.

File: _posts/2017-11-30-totw-130.md (hunk 5 above)
Comment (manshreck) [reply _posts/2017-11-30-totw-130.md]: :(

File: _posts/2017-11-30-totw-130.md (hunk 7 above)
Comment (manshreck) [reply _posts/2017-11-30-totw-130.md]: I think you have to spell out what you are "differing" (Java and C++) and what is the "mistake" (C++).

I.e.:

"As it turns out, this difference in how parent packages/namespaces are handled within Java and C++ is fundamental to why structural namespace naming (making the namespace structure match the package hierarchy) is a mistake within C++."

Comment (tituswinters) [reply _posts/2017-11-30-totw-130.md]: Done.

**Pull request merged**